src/cpp/ext/otel/otel_plugin.cc

```diff
@@ -536,6 +627,209 @@ void OpenTelemetryPlugin::RecordHistogram(
+          observable_state->instrument->RemoveCallback(
+              &ObservableCallback<int64_t>, observable_state.get());
+        },
+        [&, this](const grpc_core::GlobalInstrumentsRegistry::
+                      GlobalCallbackDoubleGaugeHandle& handle) {
+          const auto& instrument_data = instruments_data_.at(handle.index);
+          if (absl::holds_alternative<Disabled>(instrument_data.instrument)) {
+            // This instrument is disabled.
+            return;
+          }
+          GPR_ASSERT(
+              absl::holds_alternative<std::unique_ptr<ObservableState<double>>>(
+                  instrument_data.instrument));
+          auto& observable_state =
+              absl::get<std::unique_ptr<ObservableState<double>>>(
+                  instrument_data.instrument);
+          GPR_ASSERT(observable_state->callback_registered);
+          GPR_ASSERT(observable_state->registered_metric_callback == callback);
+          observable_state->instrument->RemoveCallback(
+              &ObservableCallback<double>, observable_state.get());
+        });
+  }
+  {
+    grpc_core::MutexLock l{&mu_};
+    callback_states_.erase(callback);
+  }
+}
+
+template <typename ValueType>
+void OpenTelemetryPlugin::ObservableState<ValueType>::Observe(
+    opentelemetry::metrics::ObserverResult& result) {
+  const auto& descriptor =
+      grpc_core::GlobalInstrumentsRegistry::GetInstrumentDescriptor({id});
+  grpc_core::MutexLock l{&mu};
+  GPR_ASSERT(descriptor.label_keys.size() == label_values.size());
+  GPR_ASSERT(descriptor.optional_label_keys.size() ==
+             optional_label_values.size());
+  auto& instrument_data = parent->instruments_data_.at(id);
+  opentelemetry::nostd::get<opentelemetry::nostd::shared_ptr<
+      opentelemetry::metrics::ObserverResultT<ValueType>>>(result)
+      ->Observe(value,
+                NPCMetricsKeyValueIterable(
+                    descriptor.label_keys, label_values,
+                    descriptor.optional_label_keys, optional_label_values,
+                    instrument_data.optional_labels_bits));
+}
+
+template <typename T>
+void OpenTelemetryPlugin::ObservableCallback(
+    opentelemetry::metrics::ObserverResult result, void* arg) {
+  auto* observable_state = static_cast<ObservableState<T>*>(arg);
+  if (observable_state->registered_metric_callback == nullptr) {
+    // No batching.
+    observable_state->Observe(result);
+    return;
+  }
+  grpc_core::MutexLock l{&observable_state->parent->mu_};
+  auto iter = observable_state->parent->callback_states_.find(
+      observable_state->registered_metric_callback);
+  GPR_ASSERT(iter != observable_state->parent->callback_states_.end());
+  auto now = grpc_core::Timestamp::Now();
+  if (now - iter->second.last_update_time <
+      observable_state->registered_metric_callback->min_interval()) {
+    // Use cached value.
+    observable_state->Observe(result);
+    return;
+  }
+  // Otherwise update and use the cache.
```

Comment:
Yikes, since we are updating the timestamp and updating the cache for all instruments in a `RegisteredMetricCallback` in one OTel callback (which corresponds to only one instrument), we will need to clear all the cache cells for each `RegisteredMetricCallback` first and they may be in different `CallbackGaugeState<ValueType>`s.

I'm planning to add this helper method to OpenTelemetryPlugin, this will be invoked from an OTel callback for each a RMC which registers this instrument and are due to be updated. I don't want to add additional data structure since that would seemingly complicate the design even more.

```
void OpenTelemetryPlugin ::ClearAllCacheCells(
    grpc_core::RegisteredMetricCallback* registered_metric_callback) {
  for (const auto& handle : registered_metric_callback->metrics()) {
    grpc_core::Match(
        handle,
        [&](const grpc_core::GlobalInstrumentsRegistry::
                GlobalCallbackInt64GaugeHandle& handle) {
          auto& callback_gauge_state =
              absl::get<std::unique_ptr<CallbackGaugeState<int64_t>>>(
                  instruments_data_.at(handle.index).instrument);
          grpc_core::MutexLock lock(&callback_gauge_state->mu);
          callback_gauge_state->caches[registered_metric_callback].clear();
        },
        [&](const grpc_core::GlobalInstrumentsRegistry::
                GlobalCallbackDoubleGaugeHandle& handle) {
          auto& callback_gauge_state =
              absl::get<std::unique_ptr<CallbackGaugeState<double>>>(
                  instruments_data_.at(handle.index).instrument);
          grpc_core::MutexLock lock(&callback_gauge_state->mu);
          callback_gauge_state->caches[registered_metric_callback].clear();
        });
  }
}
```

And figuring out or simplifying some of the locking stories, this does become more complicated than I thought.

Reply:
Just kidding, I think a good place to clear all previous data for a `RegisteredMetricCallback` is in the `CallbackMetricReporter`'s constructor. Though doing it this way we are making it clear that all the old data will be gone and the caller needs to fill in the new data using `CallbackMetricReporter::Report()` for all the instruments it has registered. I think this way is clear.

Also I have removed the per-instrument mutex in `CallbackGaugeState` to simplify some of the locking stories. We now only lock the plugin's mutex in `AddCallback()`, `RemoveCallback()` and `CallbackGaugeCallback<ValueType>()`. I think this simplifies the code a lot with just a little lost in parallelism (we know that the OTel callbacks are invoked sequentially from OTel). It also reduces the calls to lock and unlock mutexes. Note that we still have per-instrument mutex in `GaugeState` since we need to synchronize between `SetGauge()` and `GaugeCallback()` and having per-instrument mutex makes it more efficient.